### PR TITLE
Added an option to retain the case of campaign parameter values via SystemSetting, #PG-2343

### PR DESCRIPTION
**Pull request by AltamashShaikh**
### Description:

Added an option to retain the case of campaign parameter values via SystemSetting
Fixes: #PG-2343, #27

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by sgiehl** [comment]
@AltamashShaikh It might be better to use a solution as I described here: https://github.com/matomo-org/plugin-MarketingCampaignsReporting/pull/65#issuecomment-842226282
The problem with your solution is, that it might slow down tracking, as it needs to read the system setting for every single tracking request to check if the setting is enabled or disabled.
It could also be an idea to use a config flag instead of a system setting. That way we could even allow to easily configure that per site.

**Review by snake14 — APPROVED** [review]
Looks good. Worked as expected when tested, including migration :+1:

**Comment by AltamashShaikh** [comment]
> @AltamashShaikh It might be better to use a solution as I described here: [#65 (comment)](https://github.com/matomo-org/plugin-MarketingCampaignsReporting/pull/65#issuecomment-842226282) The problem with your solution is, that it might slow down tracking, as it needs to read the system setting for every single tracking request to check if the setting is enabled or disabled. It could also be an idea to use a config flag instead of a system setting. That way we could even allow to easily configure that per site.

@sgiehl Changed to change the case during archiving, can you check if this look good to you ?

File: Updates/5.1.0.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public function doUpdate(Updater $updater)
+    {
+        // set the SystemSetting to false for existing installs as it will be true by default
+        $systemSettings = StaticContainer::get(SystemSettings::class);
+        $systemSettings->doNotChangeCaseOfUtmParameters->setIsWritableByCurrentUser(true);
+        $systemSettings->doNotChangeCaseOfUtmParameters->setValue(false);
+        $systemSettings->save();
+        $updater->executeMigrations(__FILE__, []);
```
Comment (sgiehl) [inline Updates/5.1.0.php]: This call is actually useless

File: lang/en.json [hunk 2]
```diff
@@ -18,6 +18,8 @@
         "Groups": "Campaign Groups",
         "Placements": "Campaign Placements",
         "CampaignIds": "Campaign Ids",
-        "CombinedSourcesMediums": "Campaign Source - Medium"
+        "CombinedSourcesMediums": "Campaign Source - Medium",
+        "DoNotChangeCaseOfUtmParametersTitle": "Retain campaign parameter capitalisation",
+        "DoNotChangeCaseOfUtmParametersDescription": "If enabled, Matomo will retain capitalisation of campaign parameters in order to support compatibility with various third party service providers, such as Google. Turning this setting off will turn all campaign parameters into lower case."
```
Comment (sgiehl) [inline lang/en.json]: Asked ChatGPT if that sentence can be improved for better understandability. Here's what it came up with. Feel free to use it, or to keep what you have:
```suggestion
        "DoNotChangeCaseOfUtmParametersDescription": "When enabled, Matomo will keep the original capitalization of campaign parameters to ensure compatibility with third-party services like Google. If disabled, all campaign parameters will be converted to lowercase."
```

File: lang/en.json [hunk 3]
```diff
@@ -18,6 +18,8 @@
         "Groups": "Campaign Groups",
         "Placements": "Campaign Placements",
         "CampaignIds": "Campaign Ids",
-        "CombinedSourcesMediums": "Campaign Source - Medium"
+        "CombinedSourcesMediums": "Campaign Source - Medium",
+        "DoNotChangeCaseOfUtmParametersTitle": "Retain campaign parameter capitalisation",
```
Comment (sgiehl) [inline lang/en.json]: "Keep" might be a more commonly used word.
```suggestion
        "DoNotChangeCaseOfUtmParametersTitle": "Keep campaign parameter capitalization",
```

**Review by sgiehl — APPROVED** [review]
Left some suggestions for possible improvements. Besides that the approach looks fine that way 👍

File: Updates/5.1.0.php (hunk 1 above)
Comment (AltamashShaikh) [reply Updates/5.1.0.php]: Removed

File: lang/en.json (hunk 2 above)
Comment (AltamashShaikh) [reply lang/en.json]: I asked Gillian and she made some tweaks to your suggestion and I have added the same :+1:

**Review by snake14 — APPROVED** [review]
Nice work! I like that it does the processing during archiving. I confirmed that I could invalidate and archive and see the setting change reflected in the report :+1:

**Pull request merged**